tensorflow-framework/src/main/java/org/tensorflow/framework/metrics/AUC.java

```diff
@@ -0,0 +1,1021 @@
+
+    if (labelWeights != null) {
+      // assert that labelWeights are non-negative.
+
+      this.labelWeights = labelWeights;
+      Op checks =
+          getTF()
+              .withSubScope("AUC")
+              .assertThat(
+                  getTF()
+                      .math
+                      .greaterEqual(
+                          labelWeights, cast(getTF(), getTF().constant(0), labelWeights.type())),
+                  Collections.singletonList(
+                      getTF().constant("All values of labelWeights must be non-negative.")));
+
+      Ops ltf =
+          getTF()
+              .withSubScope("updateState")
+              .withControlDependencies(Collections.singletonList(checks));
+
+      this.labelWeights = ltf.identity(this.labelWeights);
+    }
+
+    if (this.multiLabel) {
+      this.numLabels = null;
+    }
+  }
+
+  /**
+   * Initialize truePositives, falsePositives, trueNegatives, and falseNegatives variables, given
+   * the shape of the data.
+   *
+   * @param shape the prediction shape if called from updateState, otherwise null
+   */
+  @SuppressWarnings("unchecked")
+  private Map<ConfusionMatrixEnum, Assign<T>> build(Shape shape) {
+    Shape variableShape;
+    if (initialized) {
+      return Collections.EMPTY_MAP;
+    }
+
+    if (this.isMultiLabel()) {
+      if (shape == null) {
+        throw new IllegalArgumentException("For multiLabel, a shape must be provided");
+      }
+      if (shape.numDimensions() != 2)
+        throw new IllegalArgumentException(
+            String.format(
+                "labels must have rank=2 when multiLabel is true. Found rank %d.",
+                shape.numDimensions()));
+      this.numLabels = (int) shape.size(1);
+      variableShape = Shape.of(this.numThresholds, this.numLabels);
+    } else {
+      variableShape = Shape.of(this.numThresholds);
+    }
+
+    Zeros<T> zeros = new Zeros<>(getTF());
+    Operand<T> zero = zeros.call(getTF().constant(variableShape), type);
+    if (truePositives == null) {
+      truePositives = getTF().withName(getTruePositivesName()).variable(zero);
+      initializers.put(ConfusionMatrixEnum.TRUE_POSITIVES, getTF().assign(truePositives, zero));
+    }
+
+    if (falsePositives == null) {
+      falsePositives = getTF().withName(getFalsePositivesName()).variable(zero);
+      initializers.put(ConfusionMatrixEnum.FALSE_POSITIVES, getTF().assign(falsePositives, zero));
+    }
+
+    if (trueNegatives == null) {
+      trueNegatives = getTF().withName(getTrueNegativesName()).variable(zero);
+      initializers.put(ConfusionMatrixEnum.TRUE_NEGATIVES, getTF().assign(trueNegatives, zero));
+    }
+
+    if (falseNegatives == null) {
+      falseNegatives = getTF().withName(getFalseNegativesName()).variable(zero);
+      initializers.put(ConfusionMatrixEnum.FALSE_NEGATIVES, getTF().assign(falseNegatives, zero));
+    }
+
+    this.initialized = true;
+    return initializers;
+  }
+
+  /** {@inheritDoc} */
```

Comment:
Given that (if I'm understanding correctly) `AUC` doesn't support passing batched data into a single call to `updateStateList`, it would be helpful to give this method its own javadoc that explains the permitted dimensional relationships.

Reply:
There is nothing in the `AUC` or `metrics_utils.update_confusion_matrix_variables()` Python source, that does special handling of the batch dimension.  I don't see whether it is required or prohibited.

Reply:
👍 done